File: common-lib/src/main/java/edu/iuh/exchange/common/dto/ApiResponse.java
package edu.iuh.exchange.common.dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Builder;
import lombok.Getter;

import java.time.Instant;

/**
 * Chuẩn hóa format response cho toàn bộ API hệ thống.
 * Mọi response đều phải được bọc trong ApiResponse này.
 *
 * <pre>
 * {
 *   "success": true,
 *   "statusCode": 200,
 *   "message": "OK",
 *   "data": { ... },
 *   "timestamp": "2024-01-01T00:00:00Z"
 * }
 * </pre>
 */
@Getter
@Builder
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {

    private final boolean success;
    private final int statusCode;
    private final String message;
    private final T data;
    private final String timestamp;

    // ─────────────── Static Factory Methods ───────────────

    public static <T> ApiResponse<T> ok(T data) {
        return ApiResponse.<T>builder()
                .success(true)
                .statusCode(200)
                .message("Success")
                .data(data)
                .timestamp(Instant.now().toString())
                .build();
    }

    public static <T> ApiResponse<T> ok(T data, String message) {
        return ApiResponse.<T>builder()
                .success(true)
                .statusCode(200)
                .message(message)
                .data(data)
                .timestamp(Instant.now().toString())
                .build();
    }

    public static <T> ApiResponse<T> created(T data) {
        return ApiResponse.<T>builder()
                .success(true)
                .statusCode(201)
                .message("Created")
                .data(data)
                .timestamp(Instant.now().toString())
                .build();
    }

    public static <T> ApiResponse<T> error(int status, String message) {
        return ApiResponse.<T>builder()
                .success(false)
                .statusCode(status)
                .message(message)
                .timestamp(Instant.now().toString())
                .build();
    }
}

File: packages/common/src/dto/ApiResponse.js
/**
 * Chuẩn hóa format response cho toàn bộ API hệ thống.
 * Mọi response đều phải được bọc trong ApiResponse này.
 *
 * {
 *   "success": true,
 *   "statusCode": 200,
 *   "message": "OK",
 *   "data": { ... },
 *   "timestamp": "2024-01-01T00:00:00Z"
 * }
 */
export class ApiResponse {
  /**
   * @param {object} options
   * @param {boolean} options.success
   * @param {number} options.statusCode
   * @param {string} options.message
   * @param {*} [options.data]
   */
  constructor({ success, statusCode, message, data }) {
    this.success = success;
    this.statusCode = statusCode;
    this.message = message;
    if (data !== undefined) this.data = data;
    this.timestamp = new Date().toISOString();
  }

  static ok(data, message = 'Success') {
    return new ApiResponse({ success: true, statusCode: 200, message, data });
  }

  static created(data, message = 'Created') {
    return new ApiResponse({ success: true, statusCode: 201, message, data });
  }

  static error(statusCode, message) {
    return new ApiResponse({ success: false, statusCode, message });
  }
}
